apps/navstar_rest/src/navstar_rest_dns_handler.erl

```diff
@@ -0,0 +1,170 @@
+%%%-------------------------------------------------------------------
+%%% @author dgoel
+%%% @copyright (C) 2016, <COMPANY>
+%%% @doc
+%%%
+%%% @end
+%%% Created : 15. Dec 2016 3:25 PM
+%%%-------------------------------------------------------------------
+-module(navstar_rest_dns_handler).
+-author("dgoel").
+
+-include("navstar_rest.hrl").
+-include_lib("dns/include/dns.hrl").
+
+%% API
+-export([init/2]).
+-export([content_types_provided/2, allowed_methods/2, content_types_accepted/2]).
+-export([to_json/2, update_zone/2]).
+
+-define(RECORDS_FIELD, {records, riak_dt_orswot}).
+
+init(Req, Opts) ->
+    {cowboy_rest, Req, Opts}.
+
+content_types_provided(Req, State) ->
+    {[
+        {<<"application/json">>, to_json}
+    ], Req, State}.
+
+allowed_methods(Req, State) ->
+    {[<<"GET">>, <<"POST">>], Req, State}.
+
+content_types_accepted(Req, State) ->
+    {[
+        {<<"application/json">>, update_zone}
+    ], Req, State}.
+
+%% POST handler
+-spec(update_zone(Req::cowboy_req:req(), State::any()) -> {<<>>, Req2::cowboy_req:req(), State::any()}).
+update_zone(Req, State) ->
+    <<"POST">> = cowboy_req:method(Req),
+    true = cowboy_req:has_body(Req),
```

Comment:
So, cowboy is the place where I might not recommend plain old let it crash. One thing that I recommend refactoring around is cowboy_rest: https://ninenines.eu/docs/en/cowboy/2.0/guide/rest_flowcharts/ -- https://ninenines.eu/docs/en/cowboy/2.0/guide/resource_design/

It will let you split up validation, parsing, and handling into separate functions